---
title: "Coloc: a package for colocalisation analyses"
author: "Chris Wallace"
date: "`r Sys.Date()`"
output:
  rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{coloc - intro}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
<a id="org63a948f"></a>

# A brief outline of colocalisation analysis

The coloc package can be used to perform genetic colocalisation
analysis of two potentially related phenotypes, to ask whether they
share common genetic causal variant(s) in a given region.  There are a
few key references which this vignette will not duplicate (see below),
but, in brief, two approaches are implemented.

## Proportional testing

First, the proportional approach uses the fact that for two traits
sharing causal variants, regression coefficients for either trait
against any set of SNPs in the neighbourhood of those variants must be
proportional.  This test was first proposed by Plagnol et al. <sup><a id="fnr.1" class="footref" href="#fn.1">1</a></sup>
in the context of evaluating whether expression of the gene *RPS26*
mediated the association of type 1 diabetes to a region on chromosome
12q13 as had recently been proposed.  The test addressed a common
issue in genetics, and meant researchers could avoid the need to
squint at parallel manhattan plots to decide whether two traits share
causal variants.  The function `coloc.test()` in this package evolved
from code released by Vincent, but no longer available.

However, choosing **which** SNPs to use for the test is a problem.
The obvious choice is to use those most strongly associated with one
or other trait to maximise information, but doing so induces bias in
the regression coefficients, which in turn leads to increased
likelihood of spuriously rejecting the null of colocalisation, ie
a quite substantially increased type 1 error rate <sup><a id="fnr.2" class="footref" href="#fn.2">2</a></sup>.  I proposed
two alternatives to address this problem, either using a principal
component summary of genetic variation in the region to overcome the
need to select a small set of test SNPs, implemented in `coloc.pcs()`
and associated functions, or to use the ideas of Bayesian model
averaging to average p values over SNP selections, generating
posterior predictive p values, implemented in `coloc.bma()`.

See vignette: [proportionality](02_proportionality.html)

## Enumeration, under a single causal variant assumption

Proportional testing, however, requires individual level genotype
data, which are not always available.  Claudia Giambartolomei and
Vincent Plagnol proposed an alternative method, which makes use of Jon
Wakefield's work on determining approximate Bayes Factors from p
values <sup><a id="fnr.3" class="footref" href="#fn.3">3</a></sup> to
generate a enrumeration colocalisation analysis <sup><a id="fnr.4"
class="footref" href="#fn.4">4</a></sup>, implemented in the function
`coloc.summaries()`.  By assuming there is at most one causal variant
per trait, every possible configuration can be individually enumerated
and evaluated, and aggregating over these allows us to gauge the
relative support for models which support colocalisation to those that
don't.


You can see more about the enumeration approach on
[this blogpost](http://haldanessieve.org/2013/05/21/our-paper-bayesian-test-for-co-localisation-between-pairs-of-genetic-association-studies-using-summary-statistics/).

See vignette: [enumeration](03_enumeration.html)

## Sensitivity analysis

As a Bayesian method, coloc.abf() requires the user to specify prior probabilities of SSNP causality and colocalisation.  Post-hoc sensitivity analysis can be used to assess whether results are robust across a range of plausible priors.

See vignette: [sensitivity](04_sensitivity.html)

## Enumeration, relaxing the single causal variant assumption

The single variant assumption can be relaxed through conditioning.  We
have now implemented a conditioning step within coloc, which we hope
will increase use of conditioning, and proposed an alternative,
masking.

See vignette: [conditioning/masking](05_conditioning.html)

# The difference between proportional and enumeration approaches

So what are the differences between proportional and enumeration approaches?
Which should you choose?

Well, if you only have p values, then you must use enumeration.  But be aware
that a single causal variant is assumed, and that for accurate
inference, the causal variant needs to be included, so either very
dense or imputed genotyping data is needed.  The enumeration approach has
another big advantage over proportional testing: being Bayesian, it
allows you to evaluate support for each hypothesis, whereas with
proportional testing, when the null of colocalisation is not rejected,
you cannot be sure whether this reflects true colocalisation or a lack
of power.  The proportional approach is much less affected by
spareness of genotyping data, with power only slightly decreased, but
type 1 error rates unaffected.

The behaviour of the two approaches only really differs when there are
two or more causal variants.  When both are shared, proportional
testing has the same type 1 error rate, and whilst enumeration tends to still
favour \(H_4\) (shared variant), it tends to put some more weight on
\(H_3\) (distinct variants).  But when at least one is specific to one
trait and at least one is shared, their behaviour differs more
substantially.  Of course, neither approach was designed with this
case in mind, and so there is no "right" answer, but it is instructive
to understand their expected behaviour.  Proportional testing tends to
reject colocalisation, whilst enumeration tends to favour sharing.  Until the
methods are extended to incorporate this specific case, it can be
useful to compare the two approaches when complete and dense
genotyping data are available.  When the results differ, we have
tended to identify a combination of shared and distinct causal
variants.  The enumeration approach can still be applied in this case, if p
values are available conditioning on the strongest signal, as
demonstrated in our paper <sup><a id="fnr.4.100" class="footref" href="#fn.4">4</a></sup>.

# Footnotes

<sup><a id="fn.1" href="#fnr.1">1</a></sup> <http://www.ncbi.nlm.nih.gov/pubmed/19039033>

<sup><a id="fn.2" href="#fnr.2">2</a></sup> <http://onlinelibrary.wiley.com/doi/10.1002/gepi.21765/abstract>

<sup><a id="fn.3" href="#fnr.3">3</a></sup> <http://www.ncbi.nlm.nih.gov/pubmed/18642345>

<sup><a id="fn.4" href="#fnr.4">4</a></sup> <http://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1004383>

<sup><a id="fn.5" href="#fnr.5">5</a></sup> <http://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1005272>
