---
title: "Coloc: enumeration of configurations under a single causal variant assumption"
author: "Chris Wallace"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{coloc - enumeration}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
```{r sim, child="simdata.Rmd" }
```

# Fine mapping under a single causal variant assumption

The Approximate Bayes Factor colocalisation analysis described in the next section essentially works by fine mapping each trait under a single causal variant assumption and then integrating over those two posterior distributions to calculate probabilities that those variants are shared.  Of course, this means we can look at each trait on its own quite simply, and we can do that with the function `finemap.abf`.

First we analyse a single trait from our simulated data to generate summary statistics:

```{r   }
library(coloc)
Y1 <- data@df1$Y
X1 <- as.matrix(data@df1[,-1])
tests1 <- lapply(1:ncol(X1), function(i) summary(lm(Y1 ~ X1[,i]))$coefficients[2,])
p1 <- sapply(tests1,"[",4)
maf <- colMeans(X1)/2

get.beta <- function(x) {
   beta <- sapply(x,"[",1)
   varbeta <- sapply(x, "[", 2)^2
   return(list(beta=beta,varbeta=varbeta))
}
b1 <- get.beta(tests1)
```

Then we analyse them:

```{r   }
my.res <- finemap.abf(dataset=list(beta=b1$beta, varbeta=b1$varbeta, N=nrow(X1),sdY=sd(Y1),type="quant"))
head(my.res)
tail(my.res) 
```

The `SNP.PP` column shows the posterior probability that exactly that SNP is causal.  Note the last line in this data.frame does not correspond to a SNP, but to the null model, that no SNP is causal.

Finally, if you do have full genotype data as here, while this is a fast method for fine mapping, it can be sensible to consider multiple causal variant models too.  One package that allows you to do this is [GUESSFM](http://github.com/chr1swallace/GUESSFM), described in<sup><a id="fnr.5" class="footref" href="#fn.5">5</a></sup> 


<a id="org6973f37"></a>

# (Approximate) Bayes Factor colocalisation analyses


<a id="org192e16b"></a>

## Introduction

The idea behind the ABF analysis is that the association of
each trait with SNPs in a region may be summarised by a vector of 0s
and at most a single 1, with the 1 indicating the causal SNP (so,
assuming a single causal SNP for each trait).  The posterior
probability of each possible configuration can be calculated and so,
crucially, can the posterior probabilities that the traits share
their configurations.  This allows us to estimate the support for the
following cases:

-   \(H_0\): neither trait has a genetic association in the region
-   \(H_1\): only trait 1 has a genetic association in the region
-   \(H_2\): only trait 2 has a genetic association in the region
-   \(H_3\): both traits are associated, but with different causal variants
-   \(H_4\): both traits are associated and share a single causal variant


<a id="orgb0fa807"></a>

## The basic `coloc.abf` function

The function `coloc.abf` is ideally suited to the case when only
summary data are available, and requires, for each trait, either:

-   p values for each SNP
-   each SNP's minor allele frequency
-   sample size
-   ratio of cases:controls (if using a case-control trait)

or:

-   regression coefficients for each SNP
-   variance of these regression coefficients.

If regression coefficients and their variance are available, please
use these, but we can also approximate Bayes Factors from p values and
minor allele frequencies, although, note, such approximation can be
less accurate when imputed data are used.  NB, you can mix and match,
depending on the data available from each study.

A wrapper function is avalailable to run all these steps but we will
first generate the p values manually to give all the details.  We will
calculate the coefficients and their standard errors, and the set of p
values, to compare analysis of the two types of input data (but you
should use only one - the coefficients if available, otherwise the p
values).

```{r   }
Y1 <- data@df1$Y
Y2 <- data@df2$Y

X1 <- as.matrix(data@df1[,-1])
X2 <- as.matrix(data@df2[,-1])

tests1 <- lapply(1:ncol(X1), function(i) summary(lm(Y1 ~ X1[,i]))$coefficients[2,])
tests2 <- lapply(1:ncol(X2), function(i) summary(lm(Y2 ~ X2[,i]))$coefficients[2,])

p1 <- sapply(tests1,"[",4)
p2 <- sapply(tests2,"[",4)

maf <- colMeans(X2)/2

get.beta <- function(x) {
   beta <- sapply(x,"[",1)
   varbeta <- sapply(x, "[", 2)^2
   return(list(beta=beta,varbeta=varbeta))
}
b1 <- get.beta(tests1)
b2 <- get.beta(tests2)
```

Note that we are using the second dataset in that case to compute the
minor allele frequencies.  This is unlikely to make any significant
difference but one could have used dataset 1 instead.  It is now
possible to compute the probabilities of interest.

With coefficients, and std. deviation of Y known, we have the most
accurate inference.  Note that if you have a case-control trait, set
type="cc" and supply s, the proportion of samples in the dataset that
are cases (from which sdY can be derived), instead of giving sdY
directly.

```{r   }
my.res <- coloc.abf(dataset1=list(beta=b1$beta, varbeta=b1$varbeta, N=nrow(X1),sdY=sd(Y1),type="quant"),
                    dataset2=list(beta=b2$beta, varbeta=b2$varbeta, N=nrow(X2),sdY=sd(Y2),type="quant"),
                    MAF=maf)
print(my.res[[1]]) 
```

When std. deviation of Y is unknown, coloc will try and estimate sdY from variance of beta and MAF, but this can be noisy, particularly for small datasets, and will produce a warning:

```{r   }
my.res <- coloc.abf(dataset1=list(beta=b1$beta, varbeta=b1$varbeta, N=nrow(X1),type="quant"),
                    dataset2=list(beta=b2$beta, varbeta=b2$varbeta, N=nrow(X2),type="quant"),
                    MAF=maf)
print(my.res[[1]]) 
```

When only p values are available, we can use the form:

```{r   }
my.res <- coloc.abf(dataset1=list(pvalues=p1,N=nrow(X1),type="quant"),
                    dataset2=list(pvalues=p2,N=nrow(X2),type="quant"),
                    MAF=maf)
print(my.res[[1]]) 
```
