<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Coloc: enumeration of configurations under a single causal variant assumption • coloc</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Coloc: enumeration of configurations under a single causal variant assumption">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">coloc</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">4.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/a01_intro.html">Coloc: a package for colocalisation analyses</a>
    </li>
    <li>
      <a href="../articles/a02_proportionality.html">Coloc: proportionality testing</a>
    </li>
    <li>
      <a href="../articles/a03_enumeration.html">Coloc: enumeration of configurations under a single causal variant assumption</a>
    </li>
    <li>
      <a href="../articles/a04_sensitivity.html">Coloc: sensitivity to prior values</a>
    </li>
    <li>
      <a href="../articles/a05_conditioning.html">Coloc: relaxing the single causal variant assumption</a>
    </li>
    <li>
      <a href="../articles/simdata.html">Code to simulate a small dataset</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/chr1swallace/coloc">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Coloc: enumeration of configurations under a single causal variant assumption</h1>
                        <h4 class="author">Chris Wallace</h4>
            
            <h4 class="date">2020-03-02</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/chr1swallace/coloc/blob/master/vignettes/a03_enumeration.Rmd"><code>vignettes/a03_enumeration.Rmd</code></a></small>
      <div class="hidden name"><code>a03_enumeration.Rmd</code></div>

    </div>

    
    
<div id="simulate-a-small-dataset" class="section level2">
<h2 class="hasAnchor">
<a href="#simulate-a-small-dataset" class="anchor"></a>Simulate a small dataset</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(mvtnorm)
simx &lt;-<span class="st"> </span><span class="cf">function</span>(nsnps,nsamples,S,<span class="dt">maf=</span><span class="fl">0.1</span>) {
    mu &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="dv">0</span>,nsnps)
    rawvars &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/mvtnorm/man/Mvnorm.html">rmvnorm</a></span>(<span class="dt">n=</span>nsamples, <span class="dt">mean=</span>mu, <span class="dt">sigma=</span>S)
    pvars &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/Normal.html">pnorm</a></span>(rawvars)
    x &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/Binomial.html">qbinom</a></span>(<span class="dv">1</span><span class="op">-</span>pvars, <span class="dv">2</span>, maf)
}


sim.data &lt;-<span class="st"> </span><span class="cf">function</span>(<span class="dt">nsnps=</span><span class="dv">50</span>,<span class="dt">nsamples=</span><span class="dv">200</span>,<span class="dt">causals=</span><span class="kw"><a href="https://rdrr.io/r/base/Round.html">floor</a></span>(nsnps<span class="op">/</span><span class="dv">2</span>),<span class="dt">nsim=</span><span class="dv">1</span>) {
  <span class="kw"><a href="https://rdrr.io/r/base/cat.html">cat</a></span>(<span class="st">"Generate"</span>,nsim,<span class="st">"small sets of data</span><span class="ch">\n</span><span class="st">"</span>)
  ntotal &lt;-<span class="st"> </span>nsnps <span class="op">*</span><span class="st"> </span>nsamples <span class="op">*</span><span class="st"> </span>nsim
  S &lt;-<span class="st"> </span>(<span class="dv">1</span> <span class="op">-</span><span class="st"> </span>(<span class="kw"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/outer.html">outer</a></span>(<span class="dv">1</span><span class="op">:</span>nsnps,<span class="dv">1</span><span class="op">:</span>nsnps,<span class="st">`</span><span class="dt">-</span><span class="st">`</span>))<span class="op">/</span>nsnps))<span class="op">^</span><span class="dv">4</span>
  X1 &lt;-<span class="st"> </span><span class="kw">simx</span>(nsnps,ntotal,S)
  X2 &lt;-<span class="st"> </span><span class="kw">simx</span>(nsnps,ntotal,S)
  Y1 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span>(ntotal,<span class="kw"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span>(X1[,causals,<span class="dt">drop=</span><span class="ot">FALSE</span>]<span class="op">/</span><span class="dv">2</span>),<span class="dv">2</span>)
  Y2 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span>(ntotal,<span class="kw"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span>(X2[,causals,<span class="dt">drop=</span><span class="ot">FALSE</span>]<span class="op">/</span><span class="dv">2</span>),<span class="dv">2</span>)
  <span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(X1) &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(X2) &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="st">"s"</span>,<span class="dv">1</span><span class="op">:</span>nsnps,<span class="dt">sep=</span><span class="st">""</span>)
  df1 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(<span class="dt">Y=</span>Y1,X1)
  df2 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(<span class="dt">Y=</span>Y2,X2)
  <span class="cf">if</span>(nsim<span class="op">==</span><span class="dv">1</span>) {
    <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span>(<span class="kw">new</span>(<span class="st">"simdata"</span>,
               <span class="dt">df1=</span><span class="kw"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(df1),
               <span class="dt">df2=</span><span class="kw"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(df2)))
  } <span class="cf">else</span> {
    index &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/split.html">split</a></span>(<span class="dv">1</span><span class="op">:</span>(nsamples <span class="op">*</span><span class="st"> </span>nsim), <span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="dv">1</span><span class="op">:</span>nsim, nsamples))
    objects &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(index, <span class="cf">function</span>(i) <span class="kw">new</span>(<span class="st">"simdata"</span>, <span class="dt">df1=</span><span class="kw"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(df1[i,]),
                                             <span class="dt">df2=</span><span class="kw"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(df2[i,])))
    <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span>(objects)
  }
}

<span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">46411</span>)
data &lt;-<span class="st"> </span><span class="kw">sim.data</span>()</code></pre></div>
<pre><code>## Generate 1 small sets of data</code></pre>
</div>
<div id="fine-mapping-under-a-single-causal-variant-assumption" class="section level1">
<h1 class="hasAnchor">
<a href="#fine-mapping-under-a-single-causal-variant-assumption" class="anchor"></a>Fine mapping under a single causal variant assumption</h1>
<p>The Approximate Bayes Factor colocalisation analysis described in the next section essentially works by fine mapping each trait under a single causal variant assumption and then integrating over those two posterior distributions to calculate probabilities that those variants are shared. Of course, this means we can look at each trait on its own quite simply, and we can do that with the function <code>finemap.abf</code>.</p>
<p>First we analyse a single trait from our simulated data to generate summary statistics:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(coloc)</code></pre></div>
<pre><code>## This is the condmask *development* branch of coloc.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">Y1 &lt;-<span class="st"> </span>data<span class="op">@</span>df1<span class="op">$</span>Y
X1 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(data<span class="op">@</span>df1[,<span class="op">-</span><span class="dv">1</span>])
tests1 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(X1), <span class="cf">function</span>(i) <span class="kw"><a href="../reference/eta.html">summary</a></span>(<span class="kw"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span>(Y1 <span class="op">~</span><span class="st"> </span>X1[,i]))<span class="op">$</span>coefficients[<span class="dv">2</span>,])
p1 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(tests1,<span class="st">"["</span>,<span class="dv">4</span>)
maf &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/colSums.html">colMeans</a></span>(X1)<span class="op">/</span><span class="dv">2</span>

get.beta &lt;-<span class="st"> </span><span class="cf">function</span>(x) {
   beta &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(x,<span class="st">"["</span>,<span class="dv">1</span>)
   varbeta &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(x, <span class="st">"["</span>, <span class="dv">2</span>)<span class="op">^</span><span class="dv">2</span>
   <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">beta=</span>beta,<span class="dt">varbeta=</span>varbeta))
}
b1 &lt;-<span class="st"> </span><span class="kw">get.beta</span>(tests1)</code></pre></div>
<p>Then we analyse them:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my.res &lt;-<span class="st"> </span><span class="kw"><a href="../reference/finemap.abf.html">finemap.abf</a></span>(<span class="dt">dataset=</span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">beta=</span>b1<span class="op">$</span>beta, <span class="dt">varbeta=</span>b1<span class="op">$</span>varbeta, <span class="dt">N=</span><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(X1),<span class="dt">sdY=</span><span class="kw"><a href="https://rdrr.io/r/stats/sd.html">sd</a></span>(Y1),<span class="dt">type=</span><span class="st">"quant"</span>))
<span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(my.res)</code></pre></div>
<pre><code>##            V.         z.        r.     lABF.   snp prior       SNP.PP
## 1 0.002248351  0.5101674 0.9756883 -1.731428 SNP.1 1e-04 9.675623e-18
## 2 0.002218556  0.3250072 0.9760028 -1.813361 SNP.2 1e-04 8.914476e-18
## 3 0.002181928 -0.3514567 0.9763896 -1.812731 SNP.3 1e-04 8.920090e-18
## 4 0.002225538  0.4669162 0.9759291 -1.756994 SNP.4 1e-04 9.431394e-18
## 5 0.002161220 -0.2858021 0.9766084 -1.837804 SNP.5 1e-04 8.699221e-18
## 6 0.002167510  0.4312465 0.9765420 -1.785466 SNP.6 1e-04 9.166650e-18</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">tail</a></span>(my.res) </code></pre></div>
<pre><code>##             V.         z.        r.     lABF.    snp  prior       SNP.PP
## 46 0.002176805  0.2602080 0.9764437 -1.841125 SNP.46 0.0001 8.670377e-18
## 47 0.002122981  0.3517147 0.9770128 -1.825979 SNP.47 0.0001 8.802699e-18
## 48 0.002172939 -0.1407900 0.9764846 -1.865372 SNP.48 0.0001 8.462678e-18
## 49 0.002179537 -0.5589458 0.9764149 -1.721043 SNP.49 0.0001 9.776623e-18
## 50 0.002195664 -0.6658666 0.9762445 -1.653548 SNP.50 0.0001 1.045928e-17
## 51          NA         NA        NA  0.000000   null 0.9950 5.438155e-13</code></pre>
<p>The <code>SNP.PP</code> column shows the posterior probability that exactly that SNP is causal. Note the last line in this data.frame does not correspond to a SNP, but to the null model, that no SNP is causal.</p>
<p>Finally, if you do have full genotype data as here, while this is a fast method for fine mapping, it can be sensible to consider multiple causal variant models too. One package that allows you to do this is <a href="http://github.com/chr1swallace/GUESSFM">GUESSFM</a>, described in<sup><a id="fnr.5" class="footref" href="#fn.5">5</a></sup></p>
<p><a id="org6973f37"></a></p>
</div>
<div id="approximate-bayes-factor-colocalisation-analyses" class="section level1">
<h1 class="hasAnchor">
<a href="#approximate-bayes-factor-colocalisation-analyses" class="anchor"></a>(Approximate) Bayes Factor colocalisation analyses</h1>
<p><a id="org192e16b"></a></p>
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>The idea behind the ABF analysis is that the association of each trait with SNPs in a region may be summarised by a vector of 0s and at most a single 1, with the 1 indicating the causal SNP (so, assuming a single causal SNP for each trait). The posterior probability of each possible configuration can be calculated and so, crucially, can the posterior probabilities that the traits share their configurations. This allows us to estimate the support for the following cases:</p>
<ul>
<li>
<span class="math inline">\(H_0\)</span>: neither trait has a genetic association in the region</li>
<li>
<span class="math inline">\(H_1\)</span>: only trait 1 has a genetic association in the region</li>
<li>
<span class="math inline">\(H_2\)</span>: only trait 2 has a genetic association in the region</li>
<li>
<span class="math inline">\(H_3\)</span>: both traits are associated, but with different causal variants</li>
<li>
<span class="math inline">\(H_4\)</span>: both traits are associated and share a single causal variant</li>
</ul>
<p><a id="orgb0fa807"></a></p>
</div>
<div id="the-basic-coloc-abf-function" class="section level2">
<h2 class="hasAnchor">
<a href="#the-basic-coloc-abf-function" class="anchor"></a>The basic <code>coloc.abf</code> function</h2>
<p>The function <code>coloc.abf</code> is ideally suited to the case when only summary data are available, and requires, for each trait, either:</p>
<ul>
<li>p values for each SNP</li>
<li>each SNP’s minor allele frequency</li>
<li>sample size</li>
<li>ratio of cases:controls (if using a case-control trait)</li>
</ul>
<p>or:</p>
<ul>
<li>regression coefficients for each SNP</li>
<li>variance of these regression coefficients.</li>
</ul>
<p>If regression coefficients and their variance are available, please use these, but we can also approximate Bayes Factors from p values and minor allele frequencies, although, note, such approximation can be less accurate when imputed data are used. NB, you can mix and match, depending on the data available from each study.</p>
<p>A wrapper function is avalailable to run all these steps but we will first generate the p values manually to give all the details. We will calculate the coefficients and their standard errors, and the set of p values, to compare analysis of the two types of input data (but you should use only one - the coefficients if available, otherwise the p values).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">Y1 &lt;-<span class="st"> </span>data<span class="op">@</span>df1<span class="op">$</span>Y
Y2 &lt;-<span class="st"> </span>data<span class="op">@</span>df2<span class="op">$</span>Y

X1 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(data<span class="op">@</span>df1[,<span class="op">-</span><span class="dv">1</span>])
X2 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(data<span class="op">@</span>df2[,<span class="op">-</span><span class="dv">1</span>])

tests1 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(X1), <span class="cf">function</span>(i) <span class="kw"><a href="../reference/eta.html">summary</a></span>(<span class="kw"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span>(Y1 <span class="op">~</span><span class="st"> </span>X1[,i]))<span class="op">$</span>coefficients[<span class="dv">2</span>,])
tests2 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(X2), <span class="cf">function</span>(i) <span class="kw"><a href="../reference/eta.html">summary</a></span>(<span class="kw"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span>(Y2 <span class="op">~</span><span class="st"> </span>X2[,i]))<span class="op">$</span>coefficients[<span class="dv">2</span>,])

p1 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(tests1,<span class="st">"["</span>,<span class="dv">4</span>)
p2 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(tests2,<span class="st">"["</span>,<span class="dv">4</span>)

maf &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/colSums.html">colMeans</a></span>(X2)<span class="op">/</span><span class="dv">2</span>

get.beta &lt;-<span class="st"> </span><span class="cf">function</span>(x) {
   beta &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(x,<span class="st">"["</span>,<span class="dv">1</span>)
   varbeta &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(x, <span class="st">"["</span>, <span class="dv">2</span>)<span class="op">^</span><span class="dv">2</span>
   <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">beta=</span>beta,<span class="dt">varbeta=</span>varbeta))
}
b1 &lt;-<span class="st"> </span><span class="kw">get.beta</span>(tests1)
b2 &lt;-<span class="st"> </span><span class="kw">get.beta</span>(tests2)</code></pre></div>
<p>Note that we are using the second dataset in that case to compute the minor allele frequencies. This is unlikely to make any significant difference but one could have used dataset 1 instead. It is now possible to compute the probabilities of interest.</p>
<p>With coefficients, and std. deviation of Y known, we have the most accurate inference. Note that if you have a case-control trait, set type=“cc” and supply s, the proportion of samples in the dataset that are cases (from which sdY can be derived), instead of giving sdY directly.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my.res &lt;-<span class="st"> </span><span class="kw"><a href="../reference/coloc.abf.html">coloc.abf</a></span>(<span class="dt">dataset1=</span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">beta=</span>b1<span class="op">$</span>beta, <span class="dt">varbeta=</span>b1<span class="op">$</span>varbeta, <span class="dt">N=</span><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(X1),<span class="dt">sdY=</span><span class="kw"><a href="https://rdrr.io/r/stats/sd.html">sd</a></span>(Y1),<span class="dt">type=</span><span class="st">"quant"</span>),
                    <span class="dt">dataset2=</span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">beta=</span>b2<span class="op">$</span>beta, <span class="dt">varbeta=</span>b2<span class="op">$</span>varbeta, <span class="dt">N=</span><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(X2),<span class="dt">sdY=</span><span class="kw"><a href="https://rdrr.io/r/stats/sd.html">sd</a></span>(Y2),<span class="dt">type=</span><span class="st">"quant"</span>),
                    <span class="dt">MAF=</span>maf)</code></pre></div>
<pre><code>## PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf 
##  1.41e-37  2.59e-25  5.47e-16  2.41e-11  1.00e+00 
## [1] "PP abf for shared variant: 100%"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/print.html">print</a></span>(my.res) </code></pre></div>
<pre><code>## Coloc analysis of trait 1, trait 2</code></pre>
<pre><code>## 
## SNP Priors</code></pre>
<pre><code>##    p1    p2   p12 
## 1e-04 1e-04 1e-05</code></pre>
<pre><code>## 
## Hypothesis Priors</code></pre>
<pre><code>##         H0    H1    H2       H3    H4
##  0.9894755 0.005 0.005 2.45e-05 5e-04</code></pre>
<pre><code>## 
## Posterior</code></pre>
<pre><code>##        nsnps           H0           H1           H2           H3           H4 
## 5.000000e+01 1.413084e-37 2.585470e-25 5.465483e-16 2.410934e-11 1.000000e+00</code></pre>
<p>When std. deviation of Y is unknown, coloc will try and estimate sdY from variance of beta and MAF, but this can be noisy, particularly for small datasets, and will produce a warning:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my.res &lt;-<span class="st"> </span><span class="kw"><a href="../reference/coloc.abf.html">coloc.abf</a></span>(<span class="dt">dataset1=</span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">beta=</span>b1<span class="op">$</span>beta, <span class="dt">varbeta=</span>b1<span class="op">$</span>varbeta, <span class="dt">N=</span><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(X1),<span class="dt">type=</span><span class="st">"quant"</span>),
                    <span class="dt">dataset2=</span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">beta=</span>b2<span class="op">$</span>beta, <span class="dt">varbeta=</span>b2<span class="op">$</span>varbeta, <span class="dt">N=</span><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(X2),<span class="dt">type=</span><span class="st">"quant"</span>),
                    <span class="dt">MAF=</span>maf)</code></pre></div>
<pre><code>## Warning in sdY.est(d$varbeta, d$MAF, d$N): estimating sdY from maf and varbeta,
## please directly supply sdY if known

## Warning in sdY.est(d$varbeta, d$MAF, d$N): estimating sdY from maf and varbeta,
## please directly supply sdY if known</code></pre>
<pre><code>## PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf 
##  1.41e-37  2.58e-25  5.45e-16  2.41e-11  1.00e+00 
## [1] "PP abf for shared variant: 100%"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/print.html">print</a></span>(my.res) </code></pre></div>
<pre><code>## Coloc analysis of trait 1, trait 2</code></pre>
<pre><code>## 
## SNP Priors</code></pre>
<pre><code>##    p1    p2   p12 
## 1e-04 1e-04 1e-05</code></pre>
<pre><code>## 
## Hypothesis Priors</code></pre>
<pre><code>##         H0    H1    H2       H3    H4
##  0.9894755 0.005 0.005 2.45e-05 5e-04</code></pre>
<pre><code>## 
## Posterior</code></pre>
<pre><code>##        nsnps           H0           H1           H2           H3           H4 
## 5.000000e+01 1.408198e-37 2.582454e-25 5.452944e-16 2.405891e-11 1.000000e+00</code></pre>
<p>When only p values are available, we can use the form:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my.res &lt;-<span class="st"> </span><span class="kw"><a href="../reference/coloc.abf.html">coloc.abf</a></span>(<span class="dt">dataset1=</span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">pvalues=</span>p1,<span class="dt">N=</span><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(X1),<span class="dt">type=</span><span class="st">"quant"</span>),
                    <span class="dt">dataset2=</span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">pvalues=</span>p2,<span class="dt">N=</span><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(X2),<span class="dt">type=</span><span class="st">"quant"</span>),
                    <span class="dt">MAF=</span>maf)</code></pre></div>
<pre><code>## PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf 
##  2.36e-37  3.78e-25  6.24e-16  2.62e-11  1.00e+00 
## [1] "PP abf for shared variant: 100%"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/print.html">print</a></span>(my.res) </code></pre></div>
<pre><code>## Coloc analysis of trait 1, trait 2</code></pre>
<pre><code>## 
## SNP Priors</code></pre>
<pre><code>##    p1    p2   p12 
## 1e-04 1e-04 1e-05</code></pre>
<pre><code>## 
## Hypothesis Priors</code></pre>
<pre><code>##         H0    H1    H2       H3    H4
##  0.9894755 0.005 0.005 2.45e-05 5e-04</code></pre>
<pre><code>## 
## Posterior</code></pre>
<pre><code>##        nsnps           H0           H1           H2           H3           H4 
## 5.000000e+01 2.356833e-37 3.778986e-25 6.236681e-16 2.622332e-11 1.000000e+00</code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#simulate-a-small-dataset">Simulate a small dataset</a></li>
      <li><a href="#fine-mapping-under-a-single-causal-variant-assumption">Fine mapping under a single causal variant assumption</a></li>
      <li>
<a href="#approximate-bayes-factor-colocalisation-analyses">(Approximate) Bayes Factor colocalisation analyses</a><ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#the-basic-coloc-abf-function">The basic <code>coloc.abf</code> function</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Chris Wallace, Claudia Giambartolomei.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
