<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Coloc: relaxing the single causal variant assumption • coloc</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Coloc: relaxing the single causal variant assumption">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">coloc</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">4.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/a01_intro.html">Coloc: a package for colocalisation analyses</a>
    </li>
    <li>
      <a href="../articles/a02_proportionality.html">Coloc: proportionality testing</a>
    </li>
    <li>
      <a href="../articles/a03_enumeration.html">Coloc: enumeration of configurations under a single causal variant assumption</a>
    </li>
    <li>
      <a href="../articles/a04_sensitivity.html">Coloc: sensitivity to prior values</a>
    </li>
    <li>
      <a href="../articles/a05_conditioning.html">Coloc: relaxing the single causal variant assumption</a>
    </li>
    <li>
      <a href="../articles/simdata.html">Code to simulate a small dataset</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/chr1swallace/coloc">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Coloc: relaxing the single causal variant assumption</h1>
                        <h4 class="author">Chris Wallace</h4>
            
            <h4 class="date">2020-03-02</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/chr1swallace/coloc/blob/master/vignettes/a05_conditioning.Rmd"><code>vignettes/a05_conditioning.Rmd</code></a></small>
      <div class="hidden name"><code>a05_conditioning.Rmd</code></div>

    </div>

    
    
<div id="simulate-a-small-dataset" class="section level2">
<h2 class="hasAnchor">
<a href="#simulate-a-small-dataset" class="anchor"></a>Simulate a small dataset</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(mvtnorm)
simx &lt;-<span class="st"> </span><span class="cf">function</span>(nsnps,nsamples,S,<span class="dt">maf=</span><span class="fl">0.1</span>) {
    mu &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="dv">0</span>,nsnps)
    rawvars &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/mvtnorm/man/Mvnorm.html">rmvnorm</a></span>(<span class="dt">n=</span>nsamples, <span class="dt">mean=</span>mu, <span class="dt">sigma=</span>S)
    pvars &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/Normal.html">pnorm</a></span>(rawvars)
    x &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/Binomial.html">qbinom</a></span>(<span class="dv">1</span><span class="op">-</span>pvars, <span class="dv">2</span>, maf)
}


sim.data &lt;-<span class="st"> </span><span class="cf">function</span>(<span class="dt">nsnps=</span><span class="dv">50</span>,<span class="dt">nsamples=</span><span class="dv">200</span>,<span class="dt">causals=</span><span class="kw"><a href="https://rdrr.io/r/base/Round.html">floor</a></span>(nsnps<span class="op">/</span><span class="dv">2</span>),<span class="dt">nsim=</span><span class="dv">1</span>) {
  <span class="kw"><a href="https://rdrr.io/r/base/cat.html">cat</a></span>(<span class="st">"Generate"</span>,nsim,<span class="st">"small sets of data</span><span class="ch">\n</span><span class="st">"</span>)
  ntotal &lt;-<span class="st"> </span>nsnps <span class="op">*</span><span class="st"> </span>nsamples <span class="op">*</span><span class="st"> </span>nsim
  S &lt;-<span class="st"> </span>(<span class="dv">1</span> <span class="op">-</span><span class="st"> </span>(<span class="kw"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/outer.html">outer</a></span>(<span class="dv">1</span><span class="op">:</span>nsnps,<span class="dv">1</span><span class="op">:</span>nsnps,<span class="st">`</span><span class="dt">-</span><span class="st">`</span>))<span class="op">/</span>nsnps))<span class="op">^</span><span class="dv">4</span>
  X1 &lt;-<span class="st"> </span><span class="kw">simx</span>(nsnps,ntotal,S)
  X2 &lt;-<span class="st"> </span><span class="kw">simx</span>(nsnps,ntotal,S)
  Y1 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span>(ntotal,<span class="kw"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span>(X1[,causals,<span class="dt">drop=</span><span class="ot">FALSE</span>]<span class="op">/</span><span class="dv">2</span>),<span class="dv">2</span>)
  Y2 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span>(ntotal,<span class="kw"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span>(X2[,causals,<span class="dt">drop=</span><span class="ot">FALSE</span>]<span class="op">/</span><span class="dv">2</span>),<span class="dv">2</span>)
  <span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(X1) &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(X2) &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="st">"s"</span>,<span class="dv">1</span><span class="op">:</span>nsnps,<span class="dt">sep=</span><span class="st">""</span>)
  df1 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(<span class="dt">Y=</span>Y1,X1)
  df2 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(<span class="dt">Y=</span>Y2,X2)
  <span class="cf">if</span>(nsim<span class="op">==</span><span class="dv">1</span>) {
    <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span>(<span class="kw">new</span>(<span class="st">"simdata"</span>,
               <span class="dt">df1=</span><span class="kw"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(df1),
               <span class="dt">df2=</span><span class="kw"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(df2)))
  } <span class="cf">else</span> {
    index &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/split.html">split</a></span>(<span class="dv">1</span><span class="op">:</span>(nsamples <span class="op">*</span><span class="st"> </span>nsim), <span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="dv">1</span><span class="op">:</span>nsim, nsamples))
    objects &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(index, <span class="cf">function</span>(i) <span class="kw">new</span>(<span class="st">"simdata"</span>, <span class="dt">df1=</span><span class="kw"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(df1[i,]),
                                             <span class="dt">df2=</span><span class="kw"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(df2[i,])))
    <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span>(objects)
  }
}

<span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">46411</span>)
data &lt;-<span class="st"> </span><span class="kw">sim.data</span>()</code></pre></div>
<pre><code>## Generate 1 small sets of data</code></pre>
</div>
<div id="multiple-causal-variants" class="section level1">
<h1 class="hasAnchor">
<a href="#multiple-causal-variants" class="anchor"></a>Multiple causal variants</h1>
<p>First, we use the above-simulated data (which contains a single colocalising variant) as a template to generate an additional set of SNPs within which one variant is causal for only one of the two traits.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">Y1 &lt;-<span class="st"> </span>data<span class="op">@</span>df1<span class="op">$</span>Y
Y2 &lt;-<span class="st"> </span>data<span class="op">@</span>df2<span class="op">$</span>Y
Y3 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(data<span class="op">@</span>df2<span class="op">$</span>Y) <span class="co"># Y3 is unassociated with anything in X2</span>

X1 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(data<span class="op">@</span>df1[,<span class="op">-</span><span class="dv">1</span>])
X2 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(data<span class="op">@</span>df2[,<span class="op">-</span><span class="dv">1</span>])

tests1 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(X1), <span class="cf">function</span>(i) <span class="kw"><a href="../reference/eta.html">summary</a></span>(<span class="kw"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span>(Y1 <span class="op">~</span><span class="st"> </span>X1[,i]))<span class="op">$</span>coefficients[<span class="dv">2</span>,])
tests2 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(X2), <span class="cf">function</span>(i) <span class="kw"><a href="../reference/eta.html">summary</a></span>(<span class="kw"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span>(Y2 <span class="op">~</span><span class="st"> </span>X2[,i]))<span class="op">$</span>coefficients[<span class="dv">2</span>,])
tests3 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(X2), <span class="cf">function</span>(i) <span class="kw"><a href="../reference/eta.html">summary</a></span>(<span class="kw"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span>(Y3 <span class="op">~</span><span class="st"> </span>X2[,i]))<span class="op">$</span>coefficients[<span class="dv">2</span>,])

p1 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(tests1,<span class="st">"["</span>,<span class="dv">4</span>)
p2 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(tests2,<span class="st">"["</span>,<span class="dv">4</span>)
p3 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(tests3,<span class="st">"["</span>,<span class="dv">4</span>)

<span class="co"># we are going to double the number of SNPs so that Y1 has a causal</span>
<span class="co"># variant in each block (tests1, duplicated) and Y2 has a shared</span>
<span class="co"># causal variant in block 1 only (tests2) and not block 2 (tests3)</span>

snpnames=<span class="kw"><a href="https://rdrr.io/r/base/make.unique.html">make.unique</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(X2),<span class="dv">2</span>))
maf &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/colSums.html">colMeans</a></span>(X2)<span class="op">/</span><span class="dv">2</span>,<span class="dv">2</span>) 
<span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(maf) &lt;-<span class="st"> </span>snpnames
LD0 &lt;-<span class="st"> </span>LD1 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span>(X2)
nsnp=<span class="kw"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(X2)
<span class="kw"><a href="https://rdrr.io/r/base/dimnames.html">dimnames</a></span>(LD1)=<span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(snpnames[<span class="op">-</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span><span class="op">:</span>nsnp)],snpnames[<span class="op">-</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span><span class="op">:</span>nsnp)])
LD01=<span class="kw"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="dv">0</span>,nsnp,nsnp,<span class="dt">dimnames=</span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(snpnames[<span class="dv">1</span><span class="op">:</span>nsnp],snpnames[<span class="op">-</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span><span class="op">:</span>nsnp)]))
LD10=<span class="kw"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="dv">0</span>,nsnp,nsnp,<span class="dt">dimnames=</span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(snpnames[<span class="op">-</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span><span class="op">:</span>nsnp)],snpnames[<span class="dv">1</span><span class="op">:</span>nsnp]))

LD &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(LD0,LD01), <span class="kw"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(LD10, LD1))

get.beta &lt;-<span class="st"> </span><span class="cf">function</span>(x,nm) {
   beta &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(x,<span class="st">"["</span>,<span class="dv">1</span>)
   varbeta &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(x, <span class="st">"["</span>, <span class="dv">2</span>)<span class="op">^</span><span class="dv">2</span>
   <span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(beta) &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(varbeta) &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(LD)
   <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">beta=</span>beta,<span class="dt">varbeta=</span>varbeta))
}
b1 &lt;-<span class="st"> </span><span class="kw">get.beta</span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(tests1,tests1),<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(LD))
b2 &lt;-<span class="st"> </span><span class="kw">get.beta</span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(tests2,tests3),<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(LD))</code></pre></div>
<p>First, let us do a standard coloc (single causal variant) analysis to serve as a baseline comparison. The analysis concludes there is colocalisation, because it “sees” the SNPs on the left which are strongly associated with both traits. But it misses the SNPs on the right of the top left plot which are associated with only one trait.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(coloc)</code></pre></div>
<pre><code>## This is the condmask *development* branch of coloc.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">D1 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">beta=</span>b1<span class="op">$</span>beta,
    <span class="dt">varbeta=</span>b1<span class="op">$</span>varbeta,
    <span class="dt">N=</span><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(X1),
    <span class="dt">sdY=</span><span class="kw"><a href="https://rdrr.io/r/stats/sd.html">sd</a></span>(Y1),
    <span class="dt">type=</span><span class="st">"quant"</span>,
    <span class="dt">MAF=</span>maf,
    <span class="dt">LD=</span>LD,
    <span class="dt">snp=</span><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(b1<span class="op">$</span>beta),
    <span class="dt">position=</span><span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(b1<span class="op">$</span>beta))
D2 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">beta=</span>b2<span class="op">$</span>beta,
    <span class="dt">varbeta=</span>b2<span class="op">$</span>varbeta,
    <span class="dt">N=</span><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(X2),
    <span class="dt">sdY=</span><span class="kw"><a href="https://rdrr.io/r/stats/sd.html">sd</a></span>(Y2),
    <span class="dt">type=</span><span class="st">"quant"</span>,
    <span class="dt">MAF=</span>maf,
    <span class="dt">LD=</span>LD,
    <span class="dt">snp=</span><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(b2<span class="op">$</span>beta),
    <span class="dt">position=</span><span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(b1<span class="op">$</span>beta))
my.res &lt;-<span class="st"> </span><span class="kw"><a href="../reference/coloc.abf.html">coloc.abf</a></span>(<span class="dt">dataset1=</span>D1, <span class="dt">dataset2=</span>D2)</code></pre></div>
<pre><code>## PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf 
##  1.41e-37  5.17e-25  5.46e-16  9.99e-04  9.99e-01 
## [1] "PP abf for shared variant: 99.9%"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/class.html">class</a></span>(my.res)</code></pre></div>
<pre><code>## [1] "coloc_abf" "list"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## print.coloc_abf
my.res</code></pre></div>
<pre><code>## Coloc analysis of trait 1, trait 2</code></pre>
<pre><code>## 
## SNP Priors</code></pre>
<pre><code>##    p1    p2   p12 
## 1e-04 1e-04 1e-05</code></pre>
<pre><code>## 
## Hypothesis Priors</code></pre>
<pre><code>##        H0   H1   H2      H3    H4
##  0.978901 0.01 0.01 9.9e-05 0.001</code></pre>
<pre><code>## 
## Posterior</code></pre>
<pre><code>##        nsnps           H0           H1           H2           H3           H4 
## 1.000000e+02 1.411672e-37 5.165774e-25 5.460023e-16 9.990010e-04 9.990010e-01</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/sensitivity.html">sensitivity</a></span>(my.res,<span class="st">"H4 &gt; 0.9"</span>)</code></pre></div>
<pre><code>## Results pass decision rule H4 &gt; 0.9</code></pre>
<p><img src="a05_conditioning_files/figure-html/sens0-1.png" width="768"></p>
<p>Even though the sensitivity analysis itself looks good, the Manhattan plots suggest we are violating the assumption of a single causal variant per trait.<br>
We can formally check for evidence for secondary signals for each trait. finemap.signals() identifies SNPs which are independently associated in a dataset, and prints the snp names together with the z score. It is used within coloc.signals to identify separate signal SNPs, and run colocalisation as many times as needed to try and colocalise each pair of signals, but can also be run individually</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/finemap.signals.html">finemap.signals</a></span>(D1,<span class="dt">method=</span><span class="st">"cond"</span>)</code></pre></div>
<pre><code>##      s25    s25.1 
## 8.977636 9.325097</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/finemap.signals.html">finemap.signals</a></span>(D2,<span class="dt">method=</span><span class="st">"cond"</span>)</code></pre></div>
<pre><code>##      s25 
## 11.15903</code></pre>
<p>Note that every colocalisation conditions out every other signal except one for each trait. For that reason, trying to colocalise many signals per trait is not recommended. Instead, use pthr to set the significance (p value) required to call a signal. If you set if too low, you will capture signals that are non-significant, or too high and you will miss true signals. pthr=5e-8 would correspond to a genome-wide significance level for common variants in a European study, but we typically choose a slightly relaxed pthr=1e-6 on the basis that if there is one GW-significant signal in a region, we expect there is a greater chance for secondary signals to exist.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/finemap.signals.html">finemap.signals</a></span>(D1,<span class="dt">method=</span><span class="st">"cond"</span>,<span class="dt">pthr=</span><span class="fl">1e-20</span>) ## too small</code></pre></div>
<pre><code>## NULL</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/finemap.signals.html">finemap.signals</a></span>(D2,<span class="dt">method=</span><span class="st">"cond"</span>,<span class="dt">pthr=</span><span class="fl">0.1</span>) ## too big</code></pre></div>
<pre><code>##       s25     s28.1     s39.1 
## 11.159032  2.698084 -3.033356</code></pre>
<p>Now we can ask coloc to consider these as separate signals using the coloc.signals() function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">res &lt;-<span class="st"> </span><span class="kw"><a href="../reference/coloc.signals.html">coloc.signals</a></span>(D1,D2,<span class="dt">method=</span><span class="st">"cond"</span>,<span class="dt">p12=</span><span class="fl">1e-6</span>,<span class="dt">pthr=</span><span class="fl">1e-6</span>)</code></pre></div>
<pre><code>## PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf 
##  1.40e-36  5.12e-24  5.41e-15  9.90e-03  9.90e-01 
## [1] "PP abf for shared variant: 99%"
## PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf 
##  6.11e-36  2.59e-22  2.36e-14  1.00e+00  3.74e-17 
## [1] "PP abf for shared variant: 3.74e-15%"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">res</code></pre></div>
<pre><code>## Coloc analysis of trait 1, trait 2</code></pre>
<pre><code>## 
## SNP Priors</code></pre>
<pre><code>##    p1    p2   p12 
## 1e-04 1e-04 1e-06</code></pre>
<pre><code>## 
## Hypothesis Priors</code></pre>
<pre><code>##        H0   H1   H2      H3    H4
##  0.979801 0.01 0.01 9.9e-05 1e-04</code></pre>
<pre><code>## 
## Posterior</code></pre>
<pre><code>##    nsnps  hit1 hit2           H0           H1           H2         H3
## 1:   100   s25  s25 1.385377e-36 5.069549e-24 5.358316e-15 0.01960784
## 2:   100 s25.1  s25 6.111101e-36 2.585470e-22 2.363632e-14 1.00000000
##              H4
## 1: 9.803922e-01
## 2: 3.742970e-17</code></pre>
<p>Note that because we are doing multiple colocalisations, sensitivity() needs to know which to consider:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/sensitivity.html">sensitivity</a></span>(res,<span class="st">"H4 &gt; 0.9"</span>,<span class="dt">row=</span><span class="dv">1</span>)</code></pre></div>
<pre><code>## Results pass decision rule H4 &gt; 0.9</code></pre>
<p><img src="a05_conditioning_files/figure-html/sens-1.png" width="768"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/sensitivity.html">sensitivity</a></span>(res,<span class="st">"H4 &gt; 0.9"</span>,<span class="dt">row=</span><span class="dv">2</span>)</code></pre></div>
<pre><code>## Results fail decision rule H4 &gt; 0.9</code></pre>
<p><img src="a05_conditioning_files/figure-html/sens-2.png" width="768"></p>
<p>Because these signals are truly independent, we could also split them by using masking, which doesn’t condition at all, but restricts the search space for colocalisation to SNPs which are not in LD with any-but-one of each signal SNP. Here you also need to consider r2thr which sets the maximum <span class="math inline">\(r^2\)</span> between two SNPs for them to be considered independent.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/finemap.signals.html">finemap.signals</a></span>(D1,<span class="dt">method=</span><span class="st">"mask"</span>)</code></pre></div>
<pre><code>##      s25    s25.1 
## 8.977636 8.977636</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/finemap.signals.html">finemap.signals</a></span>(D2,<span class="dt">method=</span><span class="st">"mask"</span>)</code></pre></div>
<pre><code>##      s25 
## 11.15903</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">resm=<span class="kw"><a href="../reference/coloc.signals.html">coloc.signals</a></span>(D1,D2,<span class="dt">method=</span><span class="st">"mask"</span>,<span class="dt">p12=</span><span class="fl">1e-6</span>,<span class="dt">pthr=</span><span class="fl">1e-6</span>,<span class="dt">r2thr=</span><span class="fl">0.01</span>)</code></pre></div>
<pre><code>## PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf 
##  1.40e-36  5.12e-24  5.41e-15  9.90e-03  9.90e-01 
## [1] "PP abf for shared variant: 99%"</code></pre>
<pre><code>## 1</code></pre>
<pre><code>## dropping 0/100 : 0 + 0</code></pre>
<pre><code>## 2</code></pre>
<pre><code>## dropping 35/100 : 35 + 0</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">resm</code></pre></div>
<pre><code>## Coloc analysis of trait 1, trait 2</code></pre>
<pre><code>## 
## SNP Priors</code></pre>
<pre><code>##    p1    p2   p12 
## 1e-04 1e-04 1e-06</code></pre>
<pre><code>## 
## Hypothesis Priors</code></pre>
<pre><code>##        H0   H1   H2      H3    H4
##  0.979801 0.01 0.01 9.9e-05 1e-04</code></pre>
<pre><code>## 
## Posterior</code></pre>
<pre><code>##    nsnps  hit1 hit2           H0           H1           H2         H3
## 1:   100   s25  s25 1.385377e-36 5.069549e-24 5.358316e-15 0.01960784
## 2:   100 s25.1  s25 1.413084e-34 2.585470e-22 5.465483e-13 1.00000000
##              H4
## 1: 9.803922e-01
## 2: 6.033737e-25</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/sensitivity.html">sensitivity</a></span>(resm,<span class="st">"H4 &gt; 0.9"</span>,<span class="dt">row=</span><span class="dv">1</span>)</code></pre></div>
<pre><code>## Results pass decision rule H4 &gt; 0.9</code></pre>
<p><img src="a05_conditioning_files/figure-html/unnamed-chunk-5-1.png" width="768"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/sensitivity.html">sensitivity</a></span>(resm,<span class="st">"H4 &gt; 0.9"</span>,<span class="dt">row=</span><span class="dv">2</span>)</code></pre></div>
<pre><code>## Results fail decision rule H4 &gt; 0.9</code></pre>
<p><img src="a05_conditioning_files/figure-html/unnamed-chunk-5-2.png" width="768"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#simulate-a-small-dataset">Simulate a small dataset</a></li>
      <li><a href="#multiple-causal-variants">Multiple causal variants</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Chris Wallace, Claudia Giambartolomei.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
