<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Bayesian colocalisation analysis with detailed output — coloc.detail • coloc</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Bayesian colocalisation analysis with detailed output — coloc.detail" />

<meta property="og:description" content="Bayesian colocalisation analysis, detailed output" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">coloc</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">4.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/a01_intro.html">Coloc: a package for colocalisation analyses</a>
    </li>
    <li>
      <a href="../articles/a02_proportionality.html">Coloc: proportionality testing</a>
    </li>
    <li>
      <a href="../articles/a03_enumeration.html">Coloc: enumeration of configurations under a single causal variant assumption</a>
    </li>
    <li>
      <a href="../articles/a04_sensitivity.html">Coloc: sensitivity to prior values</a>
    </li>
    <li>
      <a href="../articles/a05_conditioning.html">Coloc: relaxing the single causal variant assumption</a>
    </li>
    <li>
      <a href="../articles/simdata.html">UNKNOWN TITLE</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/chr1swallace/coloc">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Bayesian colocalisation analysis with detailed output</h1>
    <small class="dont-index">Source: <a href='https://github.com/chr1swallace/coloc/blob/master/R/split.R'><code>R/split.R</code></a></small>
    <div class="hidden name"><code>coloc.detail.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Bayesian colocalisation analysis, detailed output</p>
    
    </div>

    <pre class="usage"><span class='fu'>coloc.detail</span>(<span class='no'>dataset1</span>, <span class='no'>dataset2</span>, <span class='kw'>MAF</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>p1</span> <span class='kw'>=</span> <span class='fl'>1e-04</span>, <span class='kw'>p2</span> <span class='kw'>=</span> <span class='fl'>1e-04</span>,
  <span class='kw'>p12</span> <span class='kw'>=</span> <span class='fl'>1e-05</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>dataset1</th>
      <td><p>a list with the following elements</p><dl class='dl-horizontal'>

  <dt>pvalues</dt><dd><p>P-values for each SNP in dataset 1</p></dd>

  <dt>N</dt><dd><p>Number of samples in dataset 1</p></dd>

  <dt>MAF</dt><dd><p>minor allele frequency of the variants</p></dd>

<dt>beta</dt><dd><p>regression coefficient for each SNP from dataset 1</p></dd>

<dt>varbeta</dt><dd><p>variance of beta</p></dd>

<dt>type</dt><dd><p>the type of data in dataset 1 - either "quant" or "cc" to denote quantitative or case-control</p></dd>

<dt>s</dt><dd><p>for a case control dataset, the proportion of samples in dataset 1 that are cases</p></dd>

 <dt>sdY</dt><dd><p>for a quantitative trait, the population standard deviation of the trait.  if not given, it can be estimated from the vectors of varbeta and MAF</p></dd>

<dt>snp</dt><dd><p>a character vector of snp ids, optional. If present, it will be used to merge dataset1 and dataset2.  Otherwise, the function assumes dataset1 and dataset2 contain results for the same SNPs in the same order.</p></dd>

</dl>

<p>Some of these items may be missing, but you must give</p><ul>
<li><p>always<code>type</code></p></li>
<li><p>if <code>type</code>=="cc"<code>s</code></p></li>
<li><p>if <code>type</code>=="quant" and <code>sdY</code> known<code>sdY</code></p></li>
<li><p>if <code>type</code>=="quant" and <code>sdY</code> unknown<code>beta</code>, <code>varbeta</code>, <code>N</code>, <code>MAF</code>
and then either</p></li>
<li><p><code>pvalues</code>, <code>MAF</code></p></li>
<li><p><code>beta</code>, <code>varbeta</code></p></li>
</ul></td>
    </tr>
    <tr>
      <th>dataset2</th>
      <td><p>as above, for dataset 2</p></td>
    </tr>
    <tr>
      <th>MAF</th>
      <td><p>Common minor allele frequency vector to be used for both dataset1 and dataset2, a shorthand for supplying the same vector as parts of both datasets</p></td>
    </tr>
    <tr>
      <th>p1</th>
      <td><p>prior probability a SNP is associated with trait 1, default 1e-4</p></td>
    </tr>
    <tr>
      <th>p2</th>
      <td><p>prior probability a SNP is associated with trait 2, default 1e-4</p></td>
    </tr>
    <tr>
      <th>p12</th>
      <td><p>prior probability a SNP is associated with both traits, default 1e-5</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a list of three <code>data.tables</code>s:</p><ul>
<li><p>summary is a vector giving the number of SNPs analysed, and the posterior probabilities of H0 (no causal variant), H1 (causal variant for trait 1 only), H2 (causal variant for trait 2 only), H3 (two distinct causal variants) and H4 (one common causal variant)</p></li>
<li><p>df is an annotated version of the input data containing log Approximate Bayes Factors and intermediate calculations, and the posterior probability SNP.PP.H4 of the SNP being causal for the shared signal</p></li>
<li><p>df3 is the same for all 2 SNP H3 models</p></li>
</ul>

    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This function replicates coloc.abf, but outputs more detail for
further processing using coloc.process</p>
<p>Intended to be called internally by coloc.signals</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='coloc.process.html'>coloc.process</a></code>, <code><a href='coloc.abf.html'>coloc.abf</a></code></p></div>
    

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#see-also">See also</a></li>
          </ul>

    <h2>Author</h2>
    <p>Chris Wallace</p>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Chris Wallace, Claudia Giambartolomei.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

